File: chapter5_DFS:BFS/음료수 얼려 먹기/5-10.cpp
/*
    Date : 2022년 9월 13일
    Author : 이푸름
    Description: 이것이 코딩 테스트다 BFS 예제
*/

#include <iostream>

using namespace std;

int n, m;
int graph[1000][1000];

// DFS로 특정 노드를 방문하고 연결된 모든 노드들도 방문
bool dfs(int x, int y) {
    // 주어진 범위를 벗어나는 경우에는 즉시 종료
    if (x <= -1 || x >=n || y <= -1 || y >= m) {
        return false;
    }
    // 현재 노드를 아직 방문하지 않았다면
    if (graph[x][y] == 0) {
        graph[x][y] = 1;
        // 상, 하, 좌, 우의 위치들도 모두 재귀적으로 호출
        dfs(x - 1, y);
        dfs(x, y - 1);
        dfs(x + 1, y);
        dfs(x, y + 1);
        return true;
    }
    return false;
}

int main() {
    cin >> n >> m;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            scanf("%1d", &graph[i][j]);
        }
    }
    // 모든 노드(위치)에 대하여 음료수 채우기
    int result = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (dfs(i, j)) {
                result += 1;
            }
        }
    }
    cout << result << '\n';
}

File: chapter5_DFS:BFS/음료수 얼려 먹기/5-10.py
'''
    Date : 2022년 9월 15일
    Author : 이푸름
    Description: 이것이 코딩 테스트다 음료수 얼려 먹기 예제
'''

n, m = map(int, input().split())

graph = []
for i in range(n):
    graph.append(list(map(int,input())))

# DFS로 특정 노드 방문한 뒤에 연결된 모든 노드들도 방문
def dfs(x, y):
    # 범위 벗어나는 경우
    if x <= -1 or x >= n or y <= -1 or y >= m:
        return False
    # 현재 노드 방문하지 않았다면
    if graph[x][y] == 0:
        graph[x][y] = 1
        dfs(x - 1, y)
        dfs(x, y - 1)
        dfs(x + 1, y)
        dfs(x, y + 1)
        return True
    return False


# 모든 노드에 대하여 음료수 채우기
result = 0
for i in range(n):
    for j in range(m):
        if dfs(i, j) == True:
            result += 1

print(result)
